### PR TITLE
feat(hapi): create api for exchange rate 

### LINKED ISSUE #13: Save transactions
- save withdraw action
- save eosio.token transfer from eden members with election level > 0

### LINKED ISSUE #31: Parse memo field to obtain TX metadata

### LINKED ISSUE #48: Create API fot exchange rate
link:
https://www.coingecko.com/en/api

- [ ] Data base migrations for exchange rate
- [ ] Create script for history rate
- [ ] Query current exchange rate for new transactions (income and expenses)

### PR DESCRIPTION
**Pull request by warnerHurtado**
### What does this PR do?

- Resolve #13  
- Resolve #31 
- Resolve #48 

### Steps to test

1. See the changes on hapi/src/services/hyperion/index and /updaters

#### CheckList

- [x] Follow proper Markdown format
- [x] The content is adequate
- [x] The content is available in both english and spanish
- [x] I Ran a spell check

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -24,6 +24,7 @@ HAPI_HASURA_ADMIN_SECRET=myadminsecretkey
 HAPI_HYPERION_API=https://eos-hyperion.edenia.cloud
 HAPI_HYPERION_START_AT=2022-04-01T00:00:00.000+00:00
 HAPI_EDEN_ELECTION_INTERVAL=7889400
+HAPI_COINS_HISTORY=https://api.coingecko.com/api/v3
```
Comment (xavier506) [inline .env.example:27]: `HAPI_RATE_HISTORY`  may be a better name

File: hapi/src/services/hyperion/index.js [hunk 2]
```diff
@@ -93,15 +96,32 @@ const getActions = async params => {
...
-    if (idEdenElection) await updater.apply({ ...action, idElection: idEdenElection.id })
+    const updater = updaters.find(item => item.type === `${action.contract}:${action.action}`)
+
+    if (!updater) continue
+
+    const idEdenElection = action.data.from === 'genesis.eden' ?
+      await edenElectionGql.get({ eden_delegate: { account: { _eq: action.data.to } } }) :
+      await edenElectionGql.get({ eden_delegate: { account: { _eq: action.data.from } } })
+    
+    if (idEdenElection) {
```
Comment (leisterfrancisco) [inline hapi/src/services/hyperion/index.js:107]: avoid if possible nested logic.

```
if (!idEdenElection) return
```

File: hapi/src/services/hyperion/updaters/eosiotoken-transfer.updater.js [hunk 3]
```diff
@@ -1,24 +1,27 @@
 const { edenTransactionGql } = require('../../../gql')
-const { memoSplit }          = require('../../../utils')
+const { updaterUtil }        = require('../../../utils')
```
Comment (leisterfrancisco) [inline hapi/src/services/hyperion/updaters/eosiotoken-transfer.updater.js:2]: why empty space?

File: hapi/src/services/hyperion/updaters/eosiotoken-transfer.updater.js [hunk 4]
```diff
@@ -1,24 +1,27 @@
...
+        date        : action.timestamp,
         description,
-        id_election: action.idElection,
-        recipient: action.data.to,
-        type: 'expense'
+        id_election : action.idElection,
+        recipient   : action.data.to,
+        type        : 'expense',
+        eos_exchange: action.eosPrice,
+        usd_total   : action.data.amount * action.eosPrice
```
Comment (leisterfrancisco) [inline hapi/src/services/hyperion/updaters/eosiotoken-transfer.updater.js:20]: which standard are you following?

File: hapi/src/services/hyperion/updaters/genesiseden-withdraw.updater.js [hunk 5]
```diff
@@ -1,20 +1,20 @@
...
+        txid        : action.transaction_id,
+        amount      : action.data.amount,
+        category    : 'income from genesis.edens',
+        date        : action.timestamp,
+        description : action.data.memo,
+        id_election : action.idElection,
+        recipient   : action.data.to,
+        type        : 'income',
+        eos_exchange: action.eosPrice,
+        usd_total   : action.data.amount * action.eosPrice
```
Comment (leisterfrancisco) [inline hapi/src/services/hyperion/updaters/genesiseden-withdraw.updater.js:17]: same here

**Review by leisterfrancisco — CHANGES_REQUESTED** [review]

File: hapi/src/services/hyperion/index.js [hunk 6]
```diff
@@ -3,10 +3,12 @@ const moment = require('moment')
...
 const hyperionStateService = require('../../gql/hyperion-state.gql')
 const { edenElectionGql } = require('../../gql')
-const { hyperionConfig } = require('../../config')
-const { hasuraUtil, axiosUtil, sleepUtil, isEdenExpense } = require('../../utils')
+const { hyperionConfig, eosConfig } = require('../../config')
+const { hasuraUtil, axiosUtil, sleepUtil } = require('../../utils')
 
 const TIME_BEFORE_IRREVERSIBILITY = 164
+let   LAST_CONSULTED_DAY          = null
+let   LAST_CONSULTED_DATA         = null
```
Comment (leisterfrancisco) [inline hapi/src/services/hyperion/index.js:11]: why empty space?

File: hapi/src/services/hyperion/index.js [hunk 7]
```diff
@@ -93,15 +96,32 @@ const getActions = async params => {
...
-    const idEdenElection = await edenElectionGql.get({
-      eden_delegate: { account: { _eq: action.data.from } },
-    })
-
-    if (idEdenElection) await updater.apply({ ...action, idElection: idEdenElection.id })
+    const updater = updaters.find(item => item.type === `${action.contract}:${action.action}`)
+
+    if (!updater) continue
+
+    const idEdenElection = action.data.from === 'genesis.eden' ?
```
Comment (leisterfrancisco) [inline hapi/src/services/hyperion/index.js:103]: change idEdenElection -> edenElectionId

**Review by leisterfrancisco — CHANGES_REQUESTED** [review]

File: hapi/src/services/hyperion/index.js (hunk 6 above)
Comment (xavier506) [reply hapi/src/services/hyperion/index.js:11]: what data is being consulted? maybe we can find a better name for this variable

File: hapi/src/services/eden-delegates.service.js [hunk 8]
```diff
@@ -25,11 +26,23 @@ const updateEdenTable = async () => {
...
 
-      if (!registeredMember) await edenDelegatesGql.save(memberData)
+      if (!registeredMember) registeredMember = await edenDelegatesGql.save(memberData) 
+
+      const electionData = {
+        id_delegate: registeredMember.id,
+        election_round: parseInt(member[0].slice(8)) + 1,
+        delegate_level: member[1].election_rank,
+      }
+
```
Comment (leisterfrancisco) [inline hapi/src/services/eden-delegates.service.js:40]: remove empty space

File: hapi/src/services/hyperion/index.js [hunk 9]
```diff
@@ -76,6 +78,7 @@ const getActions = async params => {
       }
     }
   )
+
```
Comment (leisterfrancisco) [inline hapi/src/services/hyperion/index.js:81]: remove empty space

**Review by leisterfrancisco — CHANGES_REQUESTED** [review]

**Comment by JAMares** [comment]
Requested changes fixed

**Review by xavier506 — APPROVED** [review]
LGTM

**Pull request merged**